docs/petr4spec/Petr4-spec.mdk

```diff
@@ -142,509 +149,2979 @@ p4grammar {
+|   | $\vert$ | $\stringTyp$ | $\qquad \text{strings}$ |
+|   | $\vert$ | $\voidTyp$ | $\qquad \text{void}$ |
+|   | $\vert$ | $\dontcareTyp$ | $\qquad \text{don't care}$ |
+
+**Types of P4 expressions:**
+
+|     |      |     |     |
+|~~~~:|:~~~~:|:~~~~|:~~~~|
+| $\typ$ | $::=$ | $\boolTyp$ | $\qquad \text{booleans}$ |
+|   | $\vert$ | $\errTyp$ | $\qquad \text{errors}$ |
+|   | $\vert$ | $\integerTyp$ | $\qquad \text{infinite-precision integers}$ |
+|   | $\vert$ | $\intWidthTyp \width$ | $\qquad \text{fixed-width signed integers}$ |
+|   | $\vert$ | $\bitWidthTyp \width$ | $\qquad \text{fixed-width unsigned integers}$ |
+|   | $\vert$ | $\varBitTyp \width$ | $\qquad \text{variable-width integers}$ |
+|   | $\vert$ | $\spcTyp \typ \typs$ | $\qquad \text{specialized type}$ |
+|   | $\vert$ | $\tupleTyps \typs$ | $\qquad \text{tuples}$ |
+|   | $\vert$ | $\stringTyp$ | $\qquad \text{strings}$ |
+|   | $\vert$ | $\voidTyp$ | $\qquad \text{void}$ |
+|   | $\vert$ | $\dontcareTyp$ | $\qquad \text{don't care}$ |
+|   | $\vert$ | $\arrayTyp \typ \width$ | $\qquad \text{arrays}$ |
+|   | $\vert$ | $\listTyps \typs$ | $\qquad \text{lists}$ |
+|   | $\vert$ | $\recordTyps$ | $\qquad \text{records}$ |
+|   | $\vert$ | $\setTyp \typ$ | $\qquad \text{sets}$ |
+|   | $\vert$ | $\matchKindTyp$ | $\qquad \text{match kinds}$ |
+|   | $\vert$ | $\typVar$ | $\qquad \text{type variables}$ |
+|   | $\vert$ | $\newTypeTyp \typVar \typ$ | $\qquad \text{reference}$ |
+|   | $\vert$ | $\headerTyps \field \typ$ | $\qquad \text{headers}$ |
+|   | $\vert$ | $\headerUnionTyps \field \typ $ | $\qquad \text{header unions}$ |
+|   | $\vert$ | $\structTyps \field \typ$ | $\qquad \text{structs}$ |
+|   | $\vert$ | $\enumTypDef$ | $\qquad \text{enums}$ |
+|   | $\vert$ | $\packageTyp \typParams \prms \wildcardParams$ | $\qquad \text{packages}$ |
+|   | $\vert$ | $\controlTyp \typParams \prms$ | $\qquad \text{controls}$ |
+|   | $\vert$ | $\parserTyp \typParams \prms$ | $\qquad \text{parsers}$ |
+|   | $\vert$ | $\externTyp \name$ | $\qquad \text{externs}$ |
+|   | $\vert$ | $\funcType \kind \typParams \params \typ$ | $\qquad \text{functions}$ |
+|   | $\vert$ | $\actionTyp {\overline {\prm 1}} {\overline {\prm 2}}$ | $\qquad \text{actions}$ |
+|   | $\vert$ | $\constructorTyp \typParams \prms \wildcardParams \typ$ | $\qquad \text{constructors}$ |
+|   | $\vert$ | $\tableTyp \name$ | $\qquad \text{tables}$ |
+{ .booktable }
+
+
+**Directions:**
+
+|     |      |     |     |
+|~~~~:|:~~~~:|:~~~~|:~~~~|
+| $\direction$ | $::=$ | $\inDir$ | $\qquad \text{copy-in}$ |
+|   | $\vert$ | $\out$ | $\qquad \text{copy-out}$ |
+|   | $\vert$ | $\inout$ | $\qquad \text{copy-in-out}$ |
+|   | $\vert$ | $\less$ | $\qquad \text{direction-less}$ |
+
+**Parameters:**
+
+|     |      |     |     |
+|~~~~:|:~~~~:|:~~~~|:~~~~|
+| $\prmDef$ | $::=$ | $\dir \, \var : \baseTyp$ | $\qquad \text{parameters}$ |
+|   | $\vert$ | $\dir \,\var : (\baseTyp, \exp)$ | $\qquad \text{parameters with value}$ |
```

Comment:
Syntax here also seems a bit convoluted. Why the parentheses around the type and expression? Again, should we just use P4's concrete syntax as our guide?

Reply:
applied.